pkg/localstore/gc_test.go

```diff
@@ -716,3 +717,145 @@ func addRandomChunks(t *testing.T, count int, db *DB, pin bool) []swarm.Chunk {
 	}
 	return chunks
 }
+
+// TestGC_NoEvictDirty checks that the garbage collection
+// does not evict chunks that are marked as dirty while the gc
+// is running.
+func TestGC_NoEvictDirty(t *testing.T) {
+	// lower the maximal number of chunks in a single
+	// gc batch to ensure multiple batches.
+	defer func(s uint64) { gcBatchSize = s }(gcBatchSize)
+	gcBatchSize = 2
+
+	chunkCount := 15
+
+	var closed chan struct{}
+	testHookCollectGarbageChan := make(chan uint64)
+	t.Cleanup(setTestHookCollectGarbage(func(collectedCount uint64) {
+		select {
+		case testHookCollectGarbageChan <- collectedCount:
+		case <-closed:
+		}
+	}))
+
+	dirtyChan := make(chan struct{})
+	incomingChan := make(chan struct{})
+	t.Cleanup(setTestHookGCIteratorDone(func() {
+		incomingChan <- struct{}{}
+		<-dirtyChan
+	}))
+	db := newTestDB(t, &Options{
+		Capacity: 10,
+	})
+
+	closed = db.close
+
+	addrs := make([]swarm.Address, 0)
+	mtx := new(sync.Mutex)
+	online := make(chan struct{})
+	go func() {
+		close(online) // make sure this is scheduled, otherwise test might flake
+		i := 0
+		for range incomingChan {
+			// set a chunk to be updated in gc, resulting
+			// in a removal from the gc round. but don't do this
+			// for all chunks!
+			if i < 2 {
+				mtx.Lock()
+				_, err := db.Get(context.Background(), storage.ModeGetRequest, addrs[i])
+				mtx.Unlock()
+				if err != nil {
+					t.Error(err)
+				}
+				i++
+				time.Sleep(100 * time.Millisecond)
```

Comment:
because `Get`s update the gc and this happens in a different goroutine. So we sleep to guarantee that the update happens, resulting in the `dirtyAddresses` to be updated